lcs/agents/macs/macs.py

```diff
@@ -364,111 +440,198 @@ def generalize_conditions(self,
+            #    CONFLICT = True
+            if new_cl.classifier == new_cl.parent:
+                CONFLICT = True
+
+            # Iterate all situation seen
             for p in situations_seen:
-                existing_cls = [cl for cl in population if
-                                cl.does_match(p) and cl.action == a0]
-                new_cls = [cl for cl in set_c if cl.classifier.does_match(
-                    p) and cl.classifier.action == a0]
-
-                for (existing_cl, new_cl) in itertools.product(
-                    *[existing_cls, new_cls]):
-                    if existing_cl.effect.conflicts(new_cl.classifier.effect):
-                        assert new_cl.parent is not None
-                        set_d.add(new_cl.parent)
-                    else:
-                        set_d.add(new_cl.classifier)
+                if CONFLICT:
+                    break
+                for existing_cl in existing_cls:
+                    if existing_cl.does_match(p):
+                        if new_cl.classifier.does_match(p):
+                            # assert new_cl.classifier.condition.does_match(existing_cl.condition)
+                            if new_cl.classifier.effect.conflicts(existing_cl.effect):
+                                CONFLICT = True
+
+                                break
+            if CONFLICT:
+                set_d.add(new_cl.parent)
+            else:
+                set_d.add(new_cl.classifier)
+
+        return set_d
 
-            # Analyze every possible pair of [D] for duplicates
-            # (keep most general)
-            for (c1, c2) in itertools.combinations(set_d, 2):
-                if c1.condition.is_more_general(c2.condition) and c1 in set_d:
+    def set_d_generalization(self, set_d: Set[Classifier]) -> None:
+        if len(set_d) < 2:
+            return
+
+        LDP = len(set_d)
+
+        for (c1, c2) in itertools.permutations(set_d, 2):
+            if c1 in set_d and c2 in set_d:
+                K = c1.condition.is_more_general(c2.condition)
+                if K is None:
+                    pass
+                elif K:
                     set_d.remove(c1)
-                elif c2.condition.is_more_general(
-                    c1.condition) and c2 in set_d:
+                else:
                     set_d.remove(c2)
 
-        # In the original population replace classifiers from [A]
-        # by those generated from [D]
-        for cl in set_a:
-            population.remove(cl)
-
-        for cl in set_d:
-            assert cl not in population
-            population.append(cl)
+        LD = len(set_d)
 
     def cover_transitions(self,
                           population: ClassifiersList,
                           p0: Perception,
                           a0: int,
-                          p1: Perception) -> None:
-
+                          p1: Perception,
+                          seen_situations) -> None:
```

Comment:
I guess that all `seen_situations`, `convering_session` and `new_cls` are not needed